resim/metrics/proto/metrics.proto

```diff
@@ -0,0 +1,298 @@
+// Copyright 2023 ReSim, Inc.
+//
+// Use of this source code is governed by an MIT-style
+// license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+syntax = "proto3";
+
+import "google/protobuf/timestamp.proto";
+import "resim/utils/proto/uuid.proto";
+
+package resim.metrics.proto;
+
+// -------------------
+// Data Representation
+// -------------------
+
+enum MetricsDataType {
+    NO_DATA_TYPE = 0;
+
+    DOUBLE_ARRAY_DATA_TYPE        = 1;
+    TIMESTAMP_ARRAY_DATA_TYPE     = 2;
+    UUID_ARRAY_DATA_TYPE          = 3;
+    STRING_ARRAY_DATA_TYPE        = 4;
+    METRIC_STATUS_ARRAY_DATA_TYPE = 5;
+
+    INDEXED_DOUBLE_ARRAY_DATA_TYPE        = 6;
+    INDEXED_TIMESTAMP_ARRAY_DATA_TYPE     = 7;
+    INDEXED_UUID_ARRAY_DATA_TYPE          = 8;
+    INDEXED_STRING_ARRAY_DATA_TYPE        = 9;
+    INDEXED_METRIC_STATUS_ARRAY_DATA_TYPE = 10;
+}
+
+message MetricsDataId {
+    resim.proto.UUID id = 1;
+}
+
+message DoubleArray {
+    repeated double array = 1;
+}
+
+message TimestampArray {
+    repeated google.protobuf.Timestamp array = 1;
+}
+
+message UUIDArray {
+    repeated resim.proto.UUID array = 1;
+}
+
+message StringArray {
+    repeated string array = 1;
+}
+
+message MetricStatusArray {
+    repeated MetricStatus array = 1;
+}
+
+message Array {
+    oneof array {
+        DoubleArray       doubles    = 1;
+        TimestampArray    timestamps = 2;
+        UUIDArray         uuids      = 3;
+        StringArray       strings    = 4;
+        MetricStatusArray statuses   = 5;
+    }
+}
+
+message ArrayPerCategory {
+    repeated CategorizedArray categorized_arrays = 1;
+}
+
+message CategorizedArray {
+    string category_name = 1;
+    Array  array         = 2;
+}
```

Comment:
I will do this in one final push once @iainjwhiteside's review is in, but:

- This is going to be replaced `map<string, Array>`. This much better enforces the behavior we want.